### PR TITLE
BiG-CZ Add Search Option Support, Filter CUAHSI Results

### PR DESCRIPTION
**Pull request by rajadain**
## Overview

Filters out "gridded" results, as described in the original issue, by default when fetching CUAHSI results. In cases when the gridded results may be a lot, removing them can potentially speed up the network request, thus we prefer filtering in the middleware rather than the client side. The main use case (as described [here](https://github.com/WikiWatershed/model-my-watershed/issues/2133#issuecomment-325013206)) is that users will rarely want gridded services, so we prioritize the initial request over the switching performance (the former slower and the latter faster in a client-side implementation).

To communicate complex filters such as this to the middleware, I've added an `options` query parameter, which takes a list of string flags. Their presence can enable arbitrary behavior defined on each catalog.

Currently there is only one option: `gridded` supported by the `cuahsi` catalog. If it is absent, we exclude certain services from the back-end request which delay the entire payload. If it is absent, no filtering is done.

On the front-end, we add a `has_filters` flag to the Catalog model, which shows a special filters div if present. We also add an `options` array, to store all enabled flags. For CUAHSI, this div includes an "Include Gridded" checkbox, checking which enables the `gridded` flag and fires a search request.

Connects #2133 

### Notes

The supported flags for each catalog should probably be documented somewhere. I'm not sure if this should be in a separate README, in code comments, or in a special "discovery" API endpoint. Something to think about when we add Swagger support and make the API public down the line.

The front-end implementation of the checkbox is also rather slapdash, because I had burned more time than I'd estimated on this. Since we don't know what kinds of filters we may have in the future (text, date, range, numeric, option buttons, more checkboxes, etc) so instead of attempting a generalization here, I'm deferring until we have more information.

## Testing Instructions

 * Check out this branch and run `bundle`
 * Go to [:8000/?bigcz](http://localhost:8000/?bigcz), and perform a search around Philadelphia
 * Go to the CUAHSI tab. Ensure there are no NLDAS_FORA or NLDAS_NOAH results
 * Check the "Include Gridded" box. Ensure there is a new search performed and that those results are now included
 * Ensure the other two tabs work as before

### CODE REVIEW

**Comment by caseycesari** [comment]
I'm not noticing much of a performance difference between including/not including the "gridded" services. The AoI below has 995 WDC results when "gridded" services are not included, and 1005 when they are included. Do you know if this is expected, or should the filter result in a bigger difference?

screen shot 2017-08-29 at 9 16 08 am

File: src/mmw/apps/bigcz/clients/cuahsi/search.py [hunk 1]
```diff
@@ -29,10 +29,29 @@
...
 client = Client(CATALOG_URL, timeout=settings.BIGCZ_CLIENT_TIMEOUT)
 
 
+def get_networkIDs(services, gridded=False):
+    if not gridded:
+        return [str(s.ServiceID) for s in services
+                if s.NetworkName not in GRIDDED]
+
+    # If we don't want to filter, send nothing instead of everything
+    # Sending all networkIDs returns 0 results in GetSeriesCatalogForBox2
```
Comment (caseycesari) [inline src/mmw/apps/bigcz/clients/cuahsi/search.py]: I've read this a bunch of times, and I can't quite get a handle on it. It seems like sending all networkIDs is filtering, which is why 0 results are returned?

File: src/mmw/js/src/data_catalog/models.js [hunk 2]
```diff
@@ -24,6 +24,8 @@ var Catalog = Backbone.Model.extend({
         active: false,
         results: null, // Results collection
         resultCount: 0,
+        has_filters: false, // If local filters apply
+        options: [],        // Array of String Options
```
Comment (caseycesari) [inline src/mmw/js/src/data_catalog/models.js]: Instead of adding an `options` attribute, consider subclassing the Catalog model for WDC and adding a `gridded` attribute. Historically, array attributes have [caused problems](https://stevenschobert.com/2013/01/03/arrays-and-objects-in-backbone-models) with Backbone models. I feel like the `addOption` and the `removeOptions` are somewhat duplicative of built-in Backbone methods. Also, none of the other catalogs have options yet, I'm not sure that this approach is necessary yet.

**Comment by rajadain** [comment]
> I'm not noticing much of a performance difference between including/not including the "gridded" services. I think this is because there isn't much of a difference in the number of results.

This does seem to be the case now, although I remember there being a lot more of them. Perhaps the results are different now, or Stroud has been testing with other areas of interest.

File: src/mmw/apps/bigcz/clients/cuahsi/search.py (hunk 1 above)
Comment (rajadain) [reply src/mmw/apps/bigcz/clients/cuahsi/search.py]: This comment should probably be reworded, but the behavior _is_ counterintuitive.

Assume that there are 5 networkIDs total: `[a, b, c, d, e]`. If we make a request with a subset, like `[a, b, c]` we get results with just those networkIDs. If we make a request with an empty list, `[]`, we get results with all networkIDs. If we make a request with the full list `[a, b, c, d, e]`, we get 0 results.

I will reword this to read easier.

**Comment by rajadain** [comment]
Pushed some fixups that should address the comments.

**Comment by caseycesari** [comment]
Also just a reminder to squash fixups.

**Comment by rajadain** [comment]
Thanks for taking a look! Squashed, merging now.
